Rework seccomp user notification

ManaSugi opened the pull request:
This commit reworks user notification implementation for safety.
All previous interfaces were changed.

- [x] **Investigate why the unit test sometimes fails**

Fixes: #42

Signed-off-by: Manabu Sugimoto <Manabu.Sugimoto@sony.com>

rusty-snake commented on libseccomp/src/notify.rs:
```diff
@@ -5,107 +5,299 @@
+    } else {
+        0
+    }
 }
```
```suggestion
fn get_errno() -> i32 {
    Error::last_os_error().raw_os_error().unwrap_or(0)
}
```
- private functions are not documented for library targets unless you pass `--document-private-items`

ManaSugi replied:
Thank you. I fixed it.

rusty-snake commented on libseccomp/src/notify.rs:
```diff
@@ -5,107 +5,299 @@
-        let mut req_ptr: *mut seccomp_notif = std::ptr::null_mut();
-        let mut resp_ptr: *mut seccomp_notif_resp = std::ptr::null_mut();
+/// ScmpFd represents a file-descriptor used for seccomp userspace notifications.
+type ScmpFd = i32;
```
```diff rust
+ use std::os::unix::io::RawFd;
```

```suggestion
type ScmpFd = RawFd;
```
- Just as an idea, it doesn't make any difference.
- Is this type definition private by intention? IMHO it should be `pub`lic as it is returned by functions.

ManaSugi replied:
Thank you. I forgot to define `ScmpFd` as  `pub`. I fixed it.

ManaSugi commented:
The unit test sometimes fails.

rusty-snake reviewed (commented):
I did not looked into seccomp notify or worked with it, therfore I can not say to much if the API is good / the implementation sound.

- `ScmpNorifData` and the others should have
  ```rust
  #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
  ```
  where possible.

ManaSugi commented on libseccomp/src/notify.rs:
```diff
@@ -5,107 +5,298 @@
-    pub fn get_req_args(&self) -> &[u64; 6] {
-        unsafe { &self.req.as_ref().data.args }
+impl ScmpNotifResp {
+    fn to_sys(&self, resp: *mut seccomp_notif_resp) {
```
I fixed it.

ManaSugi commented:
Solved the problem of unit tests by https://github.com/libseccomp-rs/libseccomp-rs/pull/45/commits/68c16bce27ea5d57f1d1a157e5b70eb7810e61ad

ManaSugi commented:
If there is no problem, I'll squash the commits except for the commit of `Makefile` change.

ManaSugi commented:
I squashed the commits.

ManaSugi commented:
re-measure coverage

ManaSugi commented:
Re-test with new CI

ManaSugi commented:
@rusty-snake Do you have any updates?

ManaSugi commented:
All works done, depends-on #36

ManaSugi commented:
Could you review this if you have time?

ManaSugi commented on libseccomp/src/lib.rs:
```diff
@@ -41,25 +41,34 @@
     pub micro: u32,
 }
 impl ScmpVersion {
+    pub fn new(major: u32, minor: u32, micro: u32) -> Self {
```
Is it better to create a macro instead of the `new` method?
I think it is little complicated for some users to distinguish the `new` and the `current` method by the name.

rusty-snake replied:
What about making it private? `ScmpVersion { major: 3, minor: 2, micro: 0 }` works too.

Or `fn from(version: (u32, u32, u32)) -> Self`?

ManaSugi replied:
This `new` method will be helpful for users who call `check_api` and `check_version` that are `public` function, so it is better to make it `public`. I think writing `ScmVersion { major: 3, minor: 2, micro: 0 } ` is little tiring, so if possible, I want to use the mehod like `new`.
`from` sounds good! I'll change it from the `new` to `from` if you are ok.

rusty-snake reviewed (commented):
Review for anything except `notify.rs`.

ManaSugi commented on libseccomp/src/lib.rs:
```diff
@@ -41,25 +41,34 @@
     pub micro: u32,
 }
 impl ScmpVersion {
+    pub fn new(major: u32, minor: u32, micro: u32) -> Self {
```
I fixed it to conform to #70.

ManaSugi commented on libseccomp/src/lib.rs:
```diff
@@ -842,6 +843,48 @@ impl Drop for ScmpFilterContext {
+/// check_version returns an error if the libseccomp version being used
+/// is less than the specified ScmpVersion.
+/// Argument msg is an arbitrary non-empty operation description, which
+/// is used as a part of the error message returned.
```
I fixed it to conform to #70.

rusty-snake commented on libseccomp/src/lib.rs:
```diff
@@ -61,7 +61,14 @@ pub struct ScmpVersion {
 }
 impl ScmpVersion {
-    pub fn new(major: u32, minor: u32, micro: u32) -> Self {
+    /// Creates the `ScmpVersion` from the specified arbitrary version
```
```suggestion
    /// Creates a `ScmpVersion` from the specified arbitrary version
```

ManaSugi replied:
Thanks, I fixed it.

ManaSugi commented on libseccomp/src/lib.rs:
```diff
@@ -842,6 +843,48 @@ impl Drop for ScmpFilterContext {
+/// is used as a part of the error message returned.
+pub fn check_version(msg: &str, expected: ScmpVersion) -> Result<()> {
+    let current = ScmpVersion::current()?;
+    if (current.major < expected.major)
```
Oh, That's right. The major changes cause incompatible API changes. Thank you. I fixed it.

ManaSugi commented on libseccomp/src/lib.rs:
```diff
@@ -842,6 +843,48 @@ impl Drop for ScmpFilterContext {
+        || (current.major == expected.major && current.minor < expected.minor)
+        || (current.major == expected.major
+            && current.minor == expected.minor
+            && current.micro < expected.micro)
```
Thank you for the suggestion. I applied the second way.

ManaSugi commented:
I'll fix doc comments of `notify.rs` to conform to #70 later.

ManaSugi commented:
I refactored doc comments of user notification to conform to #70.
In addtion, I changed `check_version` and `check_api` functions as follows.

* `check_version` : A public function that returns `Ok(true)` if the libseccomp version is equal to or greater than the specified version, `Ok(false)` otherwise.

* `check_api` : A public function that returns `Ok(true)` if both the libseccomp API level and the libseccomp version are equal to or greater than the specified API level and version, `Ok(false)` otherwise.

* `ensure_supported_version` : A private function that returns an error with the specified message if the libseccomp version being used is less than the specified version.

* `ensure_supported_api` : A pivate function that returns an error with the specified message if the libseccomp API level and the libseccopm version being used are less than the specified API level and version.

This is because `check_version/api` that returns `Ok(true/false)` will be useful as more general functions for users who want to 
handle an error by themselves. (previous `check_version/api` needs to take `msg` that is part of an error message. This may bother users to consider the error message).
Hence, I created them and made them `public` instead of `ensure_supported_version/api`.

@rusty-snake Could you review this again? (and Do you have any suggestions about the above functions?)

ManaSugi commented:
@rusty-snake 
I squashed the commits into 4 commits.
The changes are bellow from your last review.

* Add `check_version` and `check_api` that are public functions
* Add `ensure_supported_version` and `ensure_supported_api` that are private functions
* Change `notify_supported` from public to private.
* Refactor the doc comments in order to conform to #70 
* Set `PKG_CONFIG_ALLOW_CROSS` in `build.rs`

Could you review and approve them again if you are ok?

BTW, regarding `ScmpVersion::from` method, I'm wondering which one is better, `ScmpVersion::from(u32, u32, u32)` or `ScmpVersion::from(version: (u32, u32, u32))` as you suggested before. `from` is a method for converting the type, so `ScmpVersion::from(version: (u32, u32, u32))` seems to be good from Rust point of view (It converts `tuple`  type into `ScmpVersion`  type). However, I think `ScmpVersion::from(u32, u32, u32)` is more simple and easy to use than `ScmpVersion::from(version: (u32, u32, u32))`. Could you give me your thoughts?

rusty-snake commented:
> ScmpVersion::from

A `from()` method in Rust should convert from one type in an other, that's what you expect.

`ScmpVersion::from((5, 2, 0))` has these double `((` `))`.

Maybe `ScmpVersion::from_components(5, 2, 0)` (or parts, numbers, ...) would be better.

ManaSugi commented:
Thank you for the reply:) I wan to adopt `ScmpVersion::from((2, 5, 0))` even if it has double `(( ))`. I think some users can also use `from` as follows. What do you think?

```rust
const MINVER: (u32, u32, u32) = (2, 5, 0);

fn main() {
    check_version(ScmpVersion::from(MINVER));
}
```

rusty-snake commented:
In this case it would be even better to make `from` `const` so you can

```rust
const LIBSECCOMP_MIN_VER: ScmpVersion = ScmpVersion::from(2, 5, 0);
```

ManaSugi commented:
I changed it to `ScmpVersion::from(version: (u32, u32, u32))`.

rusty-snake commented on libseccomp/src/lib.rs:
```diff
@@ -41,25 +41,40 @@
+    ///
+    /// * `version` - A tuple that represents the version of the libseccomp library.  
+    /// The index 0, 1, and 2 represent `major`, `minor`, and `micro` respectively.
+    pub fn from(version: (u32, u32, u32)) -> Self {
```
We could move this into an `impl From<(u32, u32, u32)> for ScmpVersion`.

ManaSugi replied:
I forgot to use `impl From`. Thank you! I fixed it.
